pkg/etcd/etcd.go

```diff
@@ -963,6 +956,100 @@ func (e *ETCD) listSnapshots(ctx context.Context, snapshotDir string) ([]snapsho
 	return snapshots, nil
 }
 
+// initS3IfNil initializes the S3 client
+// if it hasn't yet been initialized.
+func (e *ETCD) initS3IfNil(ctx context.Context) error {
+	if e.s3 == nil {
+		s3, err := newS3(ctx, e.config)
+		if err != nil {
+			return err
+		}
+		e.s3 = s3
+	}
+
+	return nil
+}
+
+// deleteSnapshots removes the given snapshots from
+// either local storage or S3.
+func (e *ETCD) DeleteSnapshots(ctx context.Context, snapshots []string) error {
+	snapshotDir, err := snapshotDir(e.config)
+	if err != nil {
+		return errors.Wrap(err, "failed to get the snapshot dir")
+	}
+
+	if e.config.EtcdS3 {
+		logrus.Info("Removing the given etcd snapshot(s) from S3")
+		logrus.Debugf("Removing the given etcd snapshot(s) from S3: %v", snapshots)
+
+		if e.initS3IfNil(ctx); err != nil {
+			return err
+		}
+
+		objectsCh := make(chan minio.ObjectInfo)
+		defer close(objectsCh)
+
+		ctx, cancel := context.WithTimeout(ctx, defaultS3OpTimeout)
+		defer cancel()
+
+		go func(ctx context.Context) {
+			opts := minio.ListObjectsOptions{
+				Recursive: true,
+			}
+
+			for obj := range e.s3.client.ListObjects(ctx, e.config.EtcdS3BucketName, opts) {
+				if obj.Err != nil {
+					logrus.Error(obj.Err)
+					return
+				}
+
+				// iterate through the given snapshots and only
+				// add them to the channel for remove if they're
+				// actually found from the bucket listing.
+				for _, snapshot := range snapshots {
+					if snapshot == obj.Key {
+						objectsCh <- obj
+					}
+				}
+			}
+		}(ctx)
+
+		for {
+			select {
+			case <-ctx.Done():
+				logrus.Errorf("Unable to delete snapshot: %v", ctx.Err())
+				return e.StoreSnapshotData(ctx)
+			case <-time.After(time.Millisecond * 100):
+				continue
+			case err, ok := <-e.s3.client.RemoveObjects(ctx, e.config.EtcdS3BucketName, objectsCh, minio.RemoveObjectsOptions{}):
```

Comment:
This is all still running with the main application context, not the one with a timeout. I think it should be done inside the same 30 second timeout?

Reply:
It's running with the new context copied in at line 992.

Reply:
In that case you shouldn't need to pass ctx in to the goroutine on 995: `go func(ctx context.Context) {` - it can just use ctx from the outer scope.

Reply:
Just a Go best practice in this case is to pass when possible as outter scope can mess with it.

Reply:
right but we control all of that here right - neither inner nor outer scope are reassigning it at any point

Reply:
I understand that. Just a best practice. Regardless, updated.